bucket_test.go

```diff
@@ -464,6 +464,133 @@ func TestBucket_Delete_NonExisting(t *testing.T) {
+				srcBucket, bErr := tx.CreateBucketIfNotExists([]byte(tc.srcBucketName))
+				if bErr != nil {
+					t.Fatalf("error creating src bucket %s: %v", tc.srcBucketName, bErr)
+				}
+
+				// insert K/V pair into src bucket
+				if pErr := srcBucket.Put([]byte("foo"), []byte("0000")); pErr != nil {
+					t.Fatal(pErr)
+				}
+
+				// create subBucket within srcBucket
+				subBucket, sErr := srcBucket.CreateBucket([]byte(tc.subBucketKey))
+				if sErr != nil {
+					t.Fatal(sErr)
+				}
+				// insert K/V pair into  subBucket
+				if pErr := subBucket.Put([]byte("this is subBucket"), []byte("0000")); pErr != nil {
+					t.Fatal(pErr)
+				}
+
+				// create dst bucket
+				dstBucket, bErr := tx.CreateBucketIfNotExists([]byte(tc.dstBucketName))
+				if bErr != nil {
+					t.Fatalf("error creating dst bucket %s: %v", tc.dstBucketName, bErr)
+				}
+
+				// insert K/V pair into dst bucket
+				if pErr := dstBucket.Put([]byte("bar"), []byte("0000")); pErr != nil {
+					t.Fatal(pErr)
+				}
+
+				return nil
+			}); err != nil {
+				t.Fatal(err)
+			}
+			db.MustCheck()
+
+			// act: move the subBucket from srcBucket to dstBucket
+			if err := db.Update(func(tx *bolt.Tx) error {
+				srcBucket := tx.Bucket([]byte(tc.srcBucketName))
+				if srcBucket == nil {
+					t.Fatal(fmt.Errorf("src bucket %s does not exist: %v", tc.srcBucketName, berrors.ErrBucketNotFound))
+				}
+
+				dstBucket := tx.Bucket([]byte(tc.dstBucketName))
+				if dstBucket == nil {
+					t.Fatal(fmt.Errorf("dst bucket %s does not exist: %v", tc.dstBucketName, berrors.ErrBucketNotFound))
+				}
+
+				mvErr := tx.MoveSubBucket([]byte(tc.subBucketKey), srcBucket, dstBucket)
+				if mvErr != nil {
+					t.Fatalf("failed to move subBucket %v from srcBucket %v to dstBucket %v: %v", tc.subBucketKey, tc.srcBucketName, tc.dstBucketName, mvErr)
+				}
+
+				return nil
+			}); err != nil {
+				t.Fatal(err)
+			}
+			db.MustCheck()
+
+			// assert: check subBucket has been deleted from srcBucket, and exists in dstBucket
+			if err := db.View(func(tx *bolt.Tx) error {
+				srcBucket := tx.Bucket([]byte(tc.srcBucketName))
+				if srcBucket == nil {
+					t.Fatal(fmt.Errorf("src bucket %s does not exist: %v", tc.srcBucketName, berrors.ErrBucketNotFound))
+				}
+				curSrc := srcBucket.Cursor()
+				k, _ := curSrc.Seek([]byte(tc.subBucketKey))
+				if bytes.Equal([]byte(tc.subBucketKey), k) {
+					t.Fatal(fmt.Errorf("key %q already exists in the target bucket: %w", tc.subBucketKey, berrors.ErrIncompatibleValue))
+				}
+
+				dstBucket := tx.Bucket([]byte(tc.dstBucketName))
+				if dstBucket == nil {
+					t.Fatal(fmt.Errorf("dst bucket %s does not exist: %v", tc.dstBucketName, berrors.ErrBucketNotFound))
+				}
+				v := dstBucket.Get([]byte(tc.subBucketKey))
+				if v != nil {
+					t.Fatal()
+				}
+				subBucket := dstBucket.Bucket([]byte(tc.subBucketKey))
+				v = subBucket.Get([]byte("this is subBucket"))
+				if v == nil {
+					t.Fatal()
+				}
+				//curDst := dstBucket.Cursor()
```

Comment:
I tried to use `seek()` to verify the subBucket is moved correctly, but I got stuck :/ 

I still need to 

- check for cornet cases 
- verify that 
```
//  1. the sub-bucket cannot be found in the source bucket;
//  2. or the key already exists in the destination bucket;
//  3. the key represents a non-bucket value.
```

ptal if i am on the right path, otherwise correct me please ( fail fast ) :D

Reply:
cc @ahrtr  ^^